common/page.go

```diff
@@ -364,6 +363,127 @@ func (p *Page) initEvents() {
 	}()
 }
 
+func (p *Page) urlGroupingName(ctx context.Context, urlTag string) (string, bool) {
+	p.eventHandlersMu.RLock()
+
+	// If there are no handlers for EventConsoleAPICalled.
+	if _, ok := p.eventHandlers[EventPageMetricCalled]; !ok {
+		p.eventHandlersMu.RUnlock()
+
+		return "", false
+	}
+	p.eventHandlersMu.RUnlock()
+
+	var name string
+	var nameChanged bool
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	p.eventHandlersMu.RLock()
+
+	for _, h := range p.eventHandlers[EventPageMetricCalled] {
+		// A handler can register another handler from within itself. This is
+		// the reason to unlock the mutex before calling the handler.
+		p.eventHandlersMu.RUnlock()
+
+		nameCh := make(chan string)
+		em := &ExportedMetric{
+			ctx:    ctx,
+			urlTag: urlTag,
+			nameCh: nameCh,
+		}
+
+		// handler runs on another goroutine.
+		h(em)
+
+		// We want to wait for the name to come in from the handler. When the
+		// nameCh is closed by the handler the name is returned back upstream.
+		ok := true
+		for ok {
+			select {
+			case n, o := <-nameCh:
+				ok = o
+				if ok {
+					name = n
+					nameChanged = true
+				}
+			case <-ctx.Done():
+				return "", false
+			}
+		}
+
+		p.eventHandlersMu.RLock()
+	}
+	p.eventHandlersMu.RUnlock()
+
+	p.logger.Debugf("URLGroupingName", "name: %q nameChanged: %v", name, nameChanged)
+
+	return name, nameChanged
+}
+
+// ExportedMetric is the type that is exported to JS. It is currently only used to
+// match on the urlTag and return a name when a match is found.
+type ExportedMetric struct {
+	ctx context.Context
+
+	// The url value from the metric's url tag. It will be used to match
+	// against the url grouping regexs.
+	urlTag string
+
+	// nameCh is used to return the url tag metric name when a match is found.
+	// It is also used to help sync the call to the handler and the caller.
+	// We need to wait for the handler to complete before proceeding to return
+	// the name of the url grouping.
+	nameCh chan<- string
+}
+
+// URLGroups will contain all the url groupings.
+type URLGroups struct {
+	Groups []URLGroup
+}
+
+// URLGroup contains the single url regex and the name to give to the metric
+// if a match is found.
+type URLGroup struct {
+	// This is a regex.
+	URL string
```

Comment:
Should we name it `RegEx` then (like `URLGroup.RegEx` is very descriptive to me)?

Reply:
I prefer `url` since in the future we may add functionality to work with `glob` or even exact matches.

Reply:
Fine, but we can always change it to `URL` in the future when we need it?

Reply:
That would be a breaking change.

I think `url` is a good choice, just as k6 have done [here](https://grafana.com/docs/k6/latest/using-k6/http-requests/#url-grouping).

Reply:
Ah, got it. We export it.

Reply:
Sure, I've changed it in 1f2b350eb5aecfe2527cd9ab7880e2f1254d0df6